### PR TITLE
Support custom objects

### PR DESCRIPTION
**Pull request by guytish**

### CODE REVIEW

File: packages/marketo-adapter/src/adapter.ts [hunk 1]
```diff
@@ -73,108 +83,162 @@ export default class MarketoAdapter implements AdapterOperations {
...
+      )
+    }
+    if (isField(change) && isCustomObject(change.parent)) {
+      await this.client.removeCustomObjectField(
+        change.parent.annotations[API_NAME],
+        {
+          input: [change.name],
+        }
+      )
+      await this.approveCustomObjectIfNeeded(change.parent.annotations)
```
Comment (omrilit) [inline packages/marketo-adapter/src/adapter.ts:158]: Why should you call `approveCustomObjectIfNeeded` when a field is deleted?

File: packages/marketo-adapter/src/adapter.ts [hunk 2]
```diff
@@ -73,108 +83,162 @@ export default class MarketoAdapter implements AdapterOperations {
...
-    // )
+  private async add(change: ChangeDataType): Promise<ObjectType> {
+    if (isObjectType(change)) {
+      return this.addObjectType(change)
+    }
+    if (isField(change) && isCustomObject(change.parent)) {
+      await this.addCustomObjectField(
+        change.parent.annotations[API_NAME], { [change.name]: change },
+      )
+      await this.approveCustomObjectIfNeeded(change.parent.annotations)
```
Comment (omrilit) [inline packages/marketo-adapter/src/adapter.ts:125]: Why should you call `approveCustomObjectIfNeeded` when you add a field?

File: packages/marketo-adapter/src/adapter.ts [hunk 3]
```diff
@@ -73,108 +83,162 @@ export default class MarketoAdapter implements AdapterOperations {
...
+  }
+
+  private async updateObjectType(change: ObjectType): Promise<ObjectType> {
+    if (isCustomObject(change)) {
+      return this.updateCustomObject(change)
+    }
+    throw new Error('Can\'t update unsupported object type')
+  }
+
+  private async updateCustomObject(change: ChangeDataType): Promise<ObjectType> {
```
Comment (omrilit) [inline packages/marketo-adapter/src/adapter.ts:192]: please modify `change` type  to `ObjectType`

File: packages/marketo-adapter/src/adapter.ts [hunk 4]
```diff
@@ -73,108 +83,162 @@ export default class MarketoAdapter implements AdapterOperations {
...
+      (co.approved ? co.approved.apiName : co.draft?.apiName) === name)
+    if (filteredCustomObject.length === 1) {
+      return filteredCustomObject[0]
+    }
+    throw new Error('Too many Custom Object Types')
+  }
+
+  private async approveCustomObjectIfNeeded(annotations: Values): Promise<void> {
+    if (annotations[STATE] === APPROVED_STATE) {
+      await this.client.approveCustomObject(annotations[API_NAME])
```
Comment (omrilit) [inline packages/marketo-adapter/src/adapter.ts:241]: Not sure if it's bug or not, and I saw that you call this function few times over the code.
I would guess `approveCustomObject` should be called only if the state was **changed to** APPROVED_STATE

File: packages/marketo-adapter/src/adapter.ts (hunk 1 above)
Comment (guytish) [reply packages/marketo-adapter/src/adapter.ts:158]: If we want to approve the changes we did - we should approve it using this API.
If not, we'll need to deploy our changes and then change the status from `approvedWithDraft` to `approved` and deploy again.

File: packages/marketo-adapter/src/adapter.ts (hunk 2 above)
Comment (guytish) [reply packages/marketo-adapter/src/adapter.ts:125]: If we want to approve the changes we did - we should approve it using this API.
If not, we'll need to deploy our changes and then change the status from `approvedWithDraft` to `approved` and deploy again.

File: packages/marketo-adapter/src/adapter.ts (hunk 3 above)
Comment (guytish) [reply packages/marketo-adapter/src/adapter.ts:192]: 👍

File: packages/marketo-adapter/src/adapter.ts (hunk 4 above)
Comment (guytish) [reply packages/marketo-adapter/src/adapter.ts:241]: If we want to approve the changes we did - we should approve it using this API.
If not, we'll need to deploy our changes and then change the status from `approvedWithDraft` to `approved` and deploy again.

File: packages/marketo-adapter/src/client/client.ts [hunk 5]
```diff
@@ -75,29 +77,102 @@ export default class MarketoClient {
     typeName: string,
     _marketoMetadata: MarketoMetadata
   ): Promise<Values[]> {
-    return this.conn.create({
+    return this.conn.request({
+      method: 'POST',
       path: `/rest/v1/${typeName}/describe.json`,
```
[the remark this answers is not in the document]
Comment (guytish) [reply packages/marketo-adapter/src/client/client.ts:82]: Correct, fixed

**Review by omrilit — APPROVED** [review]

**Pull request merged**